### PR TITLE
Scoring tests and fixes

### PR DESCRIPTION
**Pull request by clusterfudge**
Addresses some components of #130, including
* Removing custom code for an itertools replacement
* Updating some docs typos
* Updating some docstrings to better reflect reality
* Fixing scoring in `Intent.validate_with_tags`, adding associated tests.

* Description of how to validate or test this PR
`./run_tests.sh` should capture new tests and linter.

* Whether you have signed a CLA (Contributor Licensing Agreement)
Signed

### CODE REVIEW

File: adapt/intent.py [hunk 1]
```diff
@@ -129,23 +135,24 @@ def validate(self, tags, confidence):
...
 
         Returns:
-            intent(intent): Resuts from validate_with_tags
+            intent(intent): Results from validate_with_tags
         """
         intent, tags = self.validate_with_tags(tags, confidence)
         return intent
 
-    def validate_with_tags(self, tags, parse_weight):
+    def validate_with_tags(self, tags, confidence):
```
Comment (clusterfudge) [inline adapt/intent.py:143]: It's unfortunate, but this actually constitutes a breaking change (specifically named arguments; caught it when merging some new tests). I have kept all the previous fixes where we were overriding confidence or using the wrong one, and we have scoring tests to validate that.

File: adapt/intent.py [hunk 2]
```diff
@@ -165,17 +171,22 @@ def validate_with_tags(self, tags, confidence):
             intent_confidence += confidence
 
         if len(self.at_least_one) > 0:
-            best_resolution = resolve_one_of(tags, self.at_least_one)
+            # we want to use a copy of the original tags here, as opposed to
+            # local_tags. This allows us to have overlaps between `required`
```
Comment (forslund) [inline adapt/intent.py]: should this be "as opposed to tags" instead of ...local_tags

**Review by forslund — APPROVED** [review]
I think the logic changes and the cleanups all look good, as far as I understand the code here atleast :)

 I added some nitpicky-comments.

Also could you drop the two commits renaming the EventEmitter class back and forth?

File: adapt/intent.py (hunk 1 above)
Comment (clusterfudge) [reply adapt/intent.py:143]: I don't think so; the parser is _so_ lightweight that I'm actually concerned about the perf of something like that. I suppose we could just emit on first use? But I don't think the changes from the PR that introduced `parse_weight` have made it too broadly into the world yet, so its uses should be minimal.

File: adapt/intent.py (hunk 2 above)
Comment (clusterfudge) [reply adapt/intent.py]: Yeah, this whole comment block was supposed to go. I'll update.

**Comment by clusterfudge** [comment]
As for commit history, I'm actually planning to squash, so the EventEmitter commits should just fall out. I'll need to add the version bump as well.

**Pull request merged**